File: code_samples/section12/example_1_linear_search/linear_search.js
const fs = require("fs");       // Provides file system operations (readFileSync)
const path = require("path");   // Provides utilities for handling file paths

// ===============================
// Linear Search (given)
// ===============================

/**
 * Performs a basic linear search on an array.
 *
 * @param {number[]} arr   Array to search
 * @param {number} target Value to locate
 *
 * @returns {number} Index of the target if found, otherwise -1
 */
function linearSearch(arr, target) {
    // Iterate through the array sequentially
    for (let i = 0; i < arr.length; i++) {

        // Check whether the current element matches the target
        if (arr[i] === target) return i;
    }

    // Target was not found in the array
    return -1;
}

// ===============================
// Linear Search with step counting
// ===============================

/**
 * Performs a linear search while counting comparison steps.
 *
 * @param {number[]} arr   Array to search
 * @param {number} target Value to locate
 *
 * @returns {{index: number, steps: number}}
 *          Object containing the index of the target (or -1 if not found)
 *          and the number of comparisons performed
 */
function linearSearchSteps(arr, target) {
    let steps = 0; // Counts how many comparisons are performed

    // Iterate through the array sequentially
    for (let i = 0; i < arr.length; i++) {
        steps++; // Count the comparison arr[i] === target

        // Check if the current element matches the target
        if (arr[i] === target) {
            return { index: i, steps }; // Return index and step count
        }
    }

    // Target was not found after scanning entire array
    return { index: -1, steps };
}

// ===============================
// Load the data file
// ===============================

/**
 * Loads integers from a text file into an array.
 *
 * The path is resolved relative to the current file (__dirname),
 * not the process working directory.
 *
 * @param {string} relativePath Relative path to the input file
 *
 * @returns {number[]} Array of integers, or an empty array on error
 */
function loadFile(relativePath) {
    // Resolve the file path relative to this script's location
    const fullPath = path.join(__dirname, relativePath);

    try {
        // Read entire file contents as UTF-8 text
        const text = fs.readFileSync(fullPath, "utf8");

        // Split on whitespace, remove empty entries, and convert to numbers
        return text
            .split(/\s+/)
            .filter(Boolean)
            .map(Number);

    } catch (err) {
        // Provide diagnostic information if file cannot be read
        console.error("Error reading:", fullPath);
        console.error("Working directory:", process.cwd());
        return []; // Return empty array to signal failure
    }
}

// ===============================
// Run tests on one dataset
// ===============================

/**
 * Executes a standard set of linear search tests on a dataset.
 *
 * @param {string} name Descriptive name of the dataset (used in output)
 * @param {number[]} arr Integer array containing the dataset
 */
function runTestsOnDataset(name, arr) {
    console.log(`=== ${name} ===`);
    console.log(`Loaded ${arr.length} integers.\n`);

    // Test 1: Search for the first element (best-case scenario)
    let r1 = linearSearchSteps(arr, arr[0]);
    console.log(`Search first (${arr[0]}): index=${r1.index}, steps=${r1.steps}`);

    // Test 2: Search for the middle element (average-case scenario)
    let midVal = arr[Math.floor(arr.length / 2)];
    let r2 = linearSearchSteps(arr, midVal);
    console.log(`Search middle (${midVal}): index=${r2.index}, steps=${r2.steps}`);

    // Test 3: Search for the last element (worst-case successful search)
    let lastVal = arr[arr.length - 1];
    let r3 = linearSearchSteps(arr, lastVal);
    console.log(`Search last (${lastVal}): index=${r3.index}, steps=${r3.steps}`);

    // Test 4: Search for a value guaranteed not to exist
    let r4 = linearSearchSteps(arr, 999999);
    console.log(`Search missing (999999): index=${r4.index}, steps=${r4.steps}`);

    console.log();
}

// ===============================
// MAIN
// ===============================

// Data file paths are resolved relative to this script's directory,
// matching the behavior of the Java implementation
const ordered = loadFile("../data/ordered.txt");
const unordered = loadFile("../data/unordered.txt");

// Validate that both datasets were loaded successfully
if (ordered.length === 0 || unordered.length === 0) {
    console.log("Missing input files-see error messages above.");
    process.exit(1); // Exit with non-zero status on failure
}

// Run linear search benchmarks on both datasets
runTestsOnDataset("Ordered Data", ordered);
runTestsOnDataset("Unordered Data", unordered);

File: code_samples/section12/example_1_linear_search/linear_search.java
package code_samples.section12.example_1_linear_search;

import java.io.*;              // Provides file I/O classes (File, FileNotFoundException)
import java.nio.file.Path;     // Represents file system paths
import java.nio.file.Paths;   // Utility class for creating Path instances
import java.util.*;            // Provides collections, Scanner, and utility classes

/**
 * Demonstrates linear search on integer datasets,
 * including step-counting to measure search cost.
 */
public class linear_search {

    /**
     * Performs a basic linear search on an integer array.
     *
     * @param arr     Array to search
     * @param target  Value to locate
     *
     * @return Index of the target if found; otherwise -1
     */
    int linearSearch(int[] arr, int target) {
        // Iterate through the array sequentially
        for (int i = 0; i < arr.length; i++) {

            // Check whether the current element matches the target
            if (arr[i] == target) {
                return i; // Return index immediately upon finding a match
            }
        }

        // Target was not found in the array
        return -1;
    }

    /**
     * Performs a linear search while counting comparison steps.
     *
     * @param arr       Array to search
     * @param target    Value to locate
     * @param outSteps  Single-element array used to return step count
     *
     * @return Index of the target if found; otherwise -1
     */
    int linearSearchCountSteps(int[] arr, int target, int[] outSteps) {
        int steps = 0; // Counts how many comparisons are made

        // Iterate through each element in order
        for (int i = 0; i < arr.length; i++) {
            steps++; // Count the comparison arr[i] == target

            // Check if current element matches the target
            if (arr[i] == target) {
                outSteps[0] = steps; // Store total steps taken
                return i;            // Return index of the found element
            }
        }

        // Target was not found after scanning entire array
        outSteps[0] = steps;         // Total comparisons equals array length
        return -1;
    }

    /**
     * Loads integers from a text file into an array.
     *
     * @param path  Path to the input file
     *
     * @return Array of integers read from the file,
     *         or an empty array if the file cannot be read
     */
    int[] loadFile(Path path) {
        List<Integer> list = new ArrayList<>(); // Temporary dynamic storage
        File file = path.toFile();               // Convert Path to File

        // Use try-with-resources to ensure the Scanner is closed automatically
        try (Scanner sc = new Scanner(file)) {

            // Read integers until no more are available
            while (sc.hasNextInt()) {
                list.add(sc.nextInt());
            }

        } catch (FileNotFoundException e) {
            // Provide diagnostic information if file cannot be found
            System.out.println("Error reading: " + file.getPath());
            System.out.println("Working directory: " + System.getProperty("user.dir"));
            return new int[0]; // Return empty array on failure
        }

        // Convert List<Integer> to primitive int[] array
        return list.stream().mapToInt(i -> i).toArray();
    }

    /**
     * Runs a standard set of linear search tests on a dataset.
     *
     * @param name  Descriptive name of the dataset (used in output)
     * @param arr   Integer array containing the dataset
     */
    void runDataset(String name, int[] arr) {
        System.out.println("=== " + name + " ===");
        System.out.println("Loaded " + arr.length + " integers\n");

        // Single-element array used to retrieve step count by reference
        int[] steps = new int[1];

        // Test 1: Search for the first element (best-case scenario)
        int idx1 = linearSearchCountSteps(arr, arr[0], steps);
        System.out.printf("Search first (%d): index=%d, steps=%d%n",
                arr[0], idx1, steps[0]);

        // Test 2: Search for the middle element (average-case scenario)
        int midVal = arr[arr.length / 2];
        int idx2 = linearSearchCountSteps(arr, midVal, steps);
        System.out.printf("Search middle (%d): index=%d, steps=%d%n",
                midVal, idx2, steps[0]);

        // Test 3: Search for the last element (worst-case successful search)
        int lastVal = arr[arr.length - 1];
        int idx3 = linearSearchCountSteps(arr, lastVal, steps);
        System.out.printf("Search last (%d): index=%d, steps=%d%n",
                lastVal, idx3, steps[0]);

        // Test 4: Search for a value guaranteed not to exist
        int idx4 = linearSearchCountSteps(arr, 999999, steps);
        System.out.printf("Search missing (999999): index=%d, steps=%d%n%n",
                idx4, steps[0]);
    }

    /**
     * Coordinates loading datasets and running search tests.
     */
    public void runTests() {
        // Expected working directory:
        //   C:\Users\smclean\Desktop\data-structures-algorithms
        // Data files live in:
        //   code_samples\section12\data

        // Build platform-independent paths to data files
        Path orderedPath   = Paths.get("code_samples", "section12", "data", "ordered.txt");
        Path unorderedPath = Paths.get("code_samples", "section12", "data", "unordered.txt");

        // Load both datasets from disk
        int[] ordered   = loadFile(orderedPath);
        int[] unordered = loadFile(unorderedPath);

        // Validate that both datasets were loaded successfully
        if (ordered.length == 0 || unordered.length == 0) {
            System.out.println("Missing input files (see error messages above for paths).");
            return; // Exit early if data could not be loaded
        }

        // Run linear search benchmarks on each dataset
        runDataset("Ordered Data", ordered);
        runDataset("Unordered Data", unordered);
    }

    /**
     * Program entry point.
     */
    public static void main(String[] args) {
        // Instantiate the class and execute test suite
        new linear_search().runTests();
    }
}
